libexec/rtld-elf/rtld_c18n.c

```diff
@@ -258,35 +271,24 @@ compart_id_to_stack_index(compart_id_t cid)
 	return (cid - offsetof(typeof(dummy), stacks) / sizeof(*dummy.stacks));
 }
 
-static void init_compart_stack(void **, compart_id_t);
+static void init_compart_stack(void *, compart_id_t);
 
 static void
-init_compart_stack(void **stk, compart_id_t cid)
+init_compart_stack(void *base, compart_id_t cid)
 {
-	/*
-	 * INVARIANT: The bottom of a compartment's stack contains a capability
-	 * to the top of the stack either when the compartment was last entered
-	 * or when it was last exited from, which ever occured later.
-	 */
-	stk[-1] = cheri_clearperm(stk - 1, CHERI_PERM_SW_VMEM);
-	if (
-#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
-	    /*
-	     * Do not check whether tracing is enabled if initialising the RTLD
-	     * stack because the global variable might not have been relocated
-	     * yet. This is only needed under the benchmark ABI, which has the
-	     * concept of an RTLD stack.
-	     */
-	    cid == C18N_RTLD_COMPART_ID ||
-#endif
-	    ld_compartment_utrace != NULL)
-		*--((uintptr_t **)stk)[-1] = cid;
+	struct stk_bottom *stk = base;
+	--stk;
+
+	*stk = (struct stk_bottom) {
```

Comment:
Can you ensure that any padding here is zeroed so that if you add the name later unwinders can determine which format is available based on the NULL address?